lib/libgcov/gcov_gcc.c

```diff
@@ -0,0 +1,411 @@
+ *   lineno_checksum: unsigned
+ *   cfg checksum: unsigned
+ *   counters: __gcov_ctr_info[number_of_gcov_counters_in_the_fn - 1]
+ */
+struct func_data {
+	/* Pointer to object containing the function */
+	const struct object_data *object;
+	/* Identifier */
+	uint32_t id;
+	/* Checksum of the line */
+	uint32_t line_chksum;
+	/* Checksum of the configuration */
+	uint32_t cfg_chksum;
+	/* Array of counters data for each type of counter */
+	struct ctr_data {
+		/* Number of counters of the type */
+		uint32_t nb_ctrs;
+		/* Array of counters value */
+		uint64_t *ctrs;
+	} ctrs_data[];
+};
+
+typedef void (*merge_fn)(uint64_t *ctrs, uint32_t nb_ctrs);
+
+/*
+ * Structure create by gcc to hold code coverade data of an object
+ *
+ * __gcov_info (func gcc/gcc/coverage.c:build_fn_info):
+ *   Version ident: unsigned
+ *   next pointer: const __gcov_info *
+ *   stamp: unsigned
+ *   Filename: const char *
+ *   merge fn array: void(*)(<int size>, unsigned)[number of gcov counters]
+ *   n_functions: unsigned
+ *   function_info pointer pointer: const
+ */
+struct object_data {
+	/* version of gcc */
+	uint32_t version;
+	/* pointer for a liked list */
+	struct object_data *next;
+	/* Date of creation */
+	uint32_t timestamp;
+	/* Name of the object */
+	const char *objectname;
+	/* Functions to merge the counters */
+	void (*merge_fn[NB_COUNTERS])(uint64_t *ctrs, uint32_t nb_ctrs);
+	/* Number of functions in the object */
+	uint32_t nb_functions;
+	/* Array of pointers on function data */
+	const struct func_data **arr_func_data;
+};
+
+/* Structure to hold data for the library */
+static struct internal_data {
+	/* Version of gcc */
+	uint32_t gcc_version;
+	/* Size of dump buffer, store the biggest size required */
+	uint32_t dump_buf_size;
+	/* Buffer to dump */
+	void *dump_buf;
+	/* List of all the objects covered */
+	struct object_data *linked_list_objects;
+	/* If a dry run is required */
+	bool dry_run_done;
+} s_int_data;
+
+/*
+ * See gcov_impl_t->get_version
+ */
+static uint32_t gcov_gcc_get_version(void)
+{
+	return s_int_data.gcc_version;
+}
+
+/*
+ * See gcov_impl_t->init
+ */
+static void gcov_gcc_init(void *cov_object_info)
+{
+	struct object_data *od = cov_object_info;
+
+	/* The objects are inserted in LIFO */
+	if (!s_int_data.linked_list_objects) {
+		s_int_data.linked_list_objects = od;
+	} else {
+		od->next = s_int_data.linked_list_objects;
+		s_int_data.linked_list_objects = od;
+	}
+
+	if (!s_int_data.gcc_version)
+		s_int_data.gcc_version = od->version;
+}
+
+/*
+ * See gcov_impl_t->merge_add
+ */
+static void gcov_gcc_merge_add(void *ctrs, uint32_t nb_ctrs)
+{
+	(void)ctrs;
+	(void)nb_ctrs;
+}
+
+static void gcov_gcc_exit(void)
+{
+	/* empty */
+}
+
+/*
+ * See gcov_impl_t->reset_all_coverage_data
+ */
+static TEE_Result gcov_gcc_reset_all_coverage_data(void)
+{
+	struct object_data *od = s_int_data.linked_list_objects;
+	const struct func_data *fd = NULL;
+	const struct ctr_data *cd = NULL;
+	uint32_t cur_fn = 0;
```

Comment:
What is the advantage of "unsigned int"? I like uint32_t because it gives precisely the range of the variable and is shorter to write.

Reply:
I'm open to define `u_int` in `<types_ext.h>` if it's too tedious to type `unsigned int`.
When you use a fixed width integer you signal that this specific width is important, often because it's supposed to fit in a hardware defined register or some ABI. I'm not sure why exactly 32 bits should be important here. When using an `unsigned int` it's often not the exact range of the type that's important, but we know that the capacity has to be large enough for what's needed though. If you suspect that a `unsigned int` isn't large enough the next choice is often `size_t`, or perhaps `unsigned long` depending on the case.